### PR TITLE
POL-53 Stranded servers, initial release

### PR DESCRIPTION
**Pull request by nshivappa**
Initial release for Report on stranded instances

### Description

Policy checks all Servers in Cloud Management and reports on any that have stranded during the boot process. It provides the option to terminate any stranded servers after approval.

### Issues Resolved

Identifies servers that have not successfully completed their boot sequence (i.e. have "stranded in booting"). Servers can optionally be terminated after approval.

### Contribution Check List

- [x] All tests pass.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable
- [x] New functionality has been documented in CHANGELOG.MD

### Running Template Link

https://governance.rightscale.com/org/1105/projects/60073/policy-incidents/5ce523bad315f500711a3c6b

### Screenshot:

image

### CODE REVIEW

File: operational/stranded_servers/Stranded_Servers.pt [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+short_description "Report and remediate any Servers that are stranded in booting. \n See the [README](https://github.com/rightscale/policy_templates/tree/master/operational/stranded_servers) and [docs.rightscale.com/policies](http://docs.rightscale.com/policies/) to learn more."
+long_description "Version: 1.0"
+category "Operational"
+severity "low"
+
+###############################################################################
+# Permissions
+###############################################################################
+
+permission "perm_read_creds" do
```
Comment (cdwilhelm) [inline operational/stranded_servers/Stranded_Servers.pt]: I don't see any reference to cred() or use the credentials.  can this be removed?

File: operational/stranded_servers/Stranded_Servers.pt [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+      "verb": "GET",
+      "host": rs_cm_host,
+      "path": "/api/instances",
+      "headers": {
+        "X-Api-Version": "1.6",
+        "X-Account": \""+rs_project_id+\""
+      },
+      "query_params": {
+        "view": "full",
+        "filter": "state=stranded in booting" + filtered_tags
```
Comment (cdwilhelm) [inline operational/stranded_servers/Stranded_Servers.pt]: the state can be just stranded as well, not just stranded in booting.

File: operational/stranded_servers/Stranded_Servers.pt [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+  code <<-EOS
+    var content=[]
+
+    for(var i=0;i<ds_list_stranded_instances.length;i++){
+      instance=ds_list_stranded_instances[i]
+	  
+	  // Constructing tags with comma separated to display in detail_template
+      var tags = instance['tags']
+      var dis_tags=""
+      for(var j=0;j<tags.length;j++){
```
Comment (cdwilhelm) [inline operational/stranded_servers/Stranded_Servers.pt]: While this works just fine, it's more efficient to use join().  

`var dis_tags=tags.join(',) `

File: operational/stranded_servers/README.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
+## Stranded Servers
+ 
+### What it does
+
+This policy checks all Servers in Cloud Management and reports on any that have stranded during the boot process. It provides the option to terminate any stranded servers after approval.
+
+### Functional Details
+ 
+The policy leverages Cloud Management APIs to identify servers that have not successfully completed their boot sequence (i.e. have "stranded in booting"). Servers can optionally be terminated after approval.
```
Comment (cdwilhelm) [inline operational/stranded_servers/README.md]: It should probably be added that locked instances will not be terminated.

**Comment by pwalwaikar** [comment]
Since Curt is handling the review, we dont need an other official review from PHX. @cvurubindi

File: operational/stranded_servers/Stranded_Servers.pt (hunk 3 above)
Comment (nshivappa) [reply operational/stranded_servers/Stranded_Servers.pt]: @cdwilhelm 
I tried that using join() but the first two letters are getting missed and all tags are displayed as a link. So I iterated and did join, manually.

image

File: operational/stranded_servers/Stranded_Servers.pt (hunk 1 above)
Comment (nshivappa) [reply operational/stranded_servers/Stranded_Servers.pt]: Made Changes.

File: operational/stranded_servers/Stranded_Servers.pt (hunk 2 above)
Comment (nshivappa) [reply operational/stranded_servers/Stranded_Servers.pt]: Thanks, I have updated.

File: operational/stranded_servers/README.md (hunk 4 above)
Comment (nshivappa) [reply operational/stranded_servers/README.md]: I have updated.

File: operational/stranded_servers/Stranded_Servers.pt (hunk 3 above)
Comment (cdwilhelm) [reply operational/stranded_servers/Stranded_Servers.pt]: that is very strange.  i've never seen that before

**Pull request merged**